src/tm-threads.c

```diff
@@ -1291,7 +1338,17 @@ static int TmThreadKillThread(ThreadVars *tv)
     }
 
     /* join it and flag it as dead */
-    pthread_join(tv->t, NULL);
+#ifdef HAVE_DPDK
+    if (tv->lcore_id != LCORE_ID_ANY && tv->type == TVT_PPT) {
+        /* wait for DPDK worker threads */
+        rte_eal_wait_lcore(tv->lcore_id);
```

Comment:
the DPDK API can only be invoked from the main thread. It can not be called from sibling thread http://doc.dpdk.org/api/rte__launch_8h.html#ae9500e1d35bd4cfb95d18c0be863cb1e

Reply:
That is true indeed. But this piece of code is always called from the main thread. Function TmThreadKillThread is called from the main thread. The function accepts ThreadVars structure of the thread to be killed. So in this case, in contrast to functions like in receive modules (e.g. receive loop), `tv` (ThreadVars structure) does not represent the thread itself but the thread it is supposed to kill. Above mentioned "if" statement then checks if the killed thread is really a DPDK worker thread. 
Attribute `tv->lcore_id` is set to a different value only if it is a DPDK thread (the structure has value LCORE_ID_ANY by default and only if the thread is spawned under DPDK, the lcore_id is assigned.) In current circumstances, only worker threads are spawned under DPDK. Other threads remain as pthreads. Therefore, checking value of `tv->type` serves as an extra check to make sure the thread is a DPDK worker thread.